File: solutions/fsharp/phone-number/PhoneNumber.fs
module PhoneNumber

open System

let validateNumber (number:string) =
    match number with
    | _ when number[0] = '0' -> Error "area code cannot start with zero"
    | _ when number[0] = '1' -> Error "area code cannot start with one"
    | _ when number[3] = '0' -> Error "exchange code cannot start with zero"
    | _ when number[3] = '1' -> Error "exchange code cannot start with one"
    | _ ->  number |> uint64 |> Ok
    
let validateLength (digits:string) =
    match digits.Length with
    | length when length > 11 -> Error "more than 11 digits"
    | 11 when digits[0] <> '1' -> Error "11 digits must start with 1"
    | 10 -> Ok digits
    | 11 -> Ok digits[1..]
    | _ -> Error "incorrect number of digits"
        
let keepDigits input = input |> Seq.filter Char.IsDigit |> String.Concat

let isPunctuation (symbol:char) = ",:;".Contains symbol
    
let validate input =
    match input with
    | _ when Seq.exists Char.IsLetter input -> Error "letters not permitted"
    | _ when Seq.exists isPunctuation input -> Error "punctuations not permitted"
    | _ -> keepDigits input |> Ok 
    
let clean input =
    validate input
    |> Result.bind validateLength
    |> Result.bind validateNumber

File: solutions/python/phone-number/phone_number.py
class PhoneNumber:
    def __init__(self, number: str):
        if any(c.isalpha() for c in number):
            raise ValueError("letters not permitted")
        if any(c in "@:!" for c in number):
            raise ValueError("punctuations not permitted")
        digits = "".join(filter(str.isdigit, number))
        if len(digits) > 11:
            raise ValueError("must not be greater than 11 digits")
        if len(digits) < 10:
            raise ValueError("must not be fewer than 10 digits")
        if len (digits) == 11 and not digits.startswith("1"):
            raise ValueError("11 digits must start with 1")
        self.number = digits[-10:]
        if self.number.startswith("0"):
            raise ValueError("area code cannot start with zero")
        if self.number.startswith("1"):
            raise ValueError("area code cannot start with one")
        exchange_code = self.number[3:6]
        if exchange_code.startswith("0"):
            raise ValueError("exchange code cannot start with zero")
        if exchange_code.startswith("1"):
            raise ValueError("exchange code cannot start with one")

    def pretty(self) -> str:
        return f"({self.area_code})-{self.number[3:6]}-{self.number[6:]}"

    @property
    def area_code(self) -> str:
        return self.number[:3]
